addons/repair/models/repair.py

```diff
@@ -246,36 +216,116 @@ def _compute_location_id(self):
+
+
+    @api.onchange('picking_type_id')
+    def _onchange_picking_type_id(self):
+        self.operation_move_ids.picking_type_id = self.picking_type_id
+
+    @api.onchange('schedule_date')
+    def _onchange_schedule_date(self):
+        self.operation_move_ids.date = self.schedule_date
+        if self.move_id:
+            self.move_id.date = self.schedule_date
+
+
     @api.ondelete(at_uninstall=False)
     def _unlink_except_confirmed(self):
-        for order in self:
-            if order.invoice_id and order.invoice_id.posted_before:
-                raise UserError(_('You can not delete a repair order which is linked to an invoice which has been posted once.'))
-            if order.state not in ('draft', 'cancel'):
-                raise UserError(_('You can not delete a repair order once it has been confirmed. You must first cancel it.'))
+        if any(order.state not in ('draft', 'cancel') for order in self):
+            raise UserError(_('You can not delete a repair order once it has been confirmed. You must first cancel it.'))
 
     @api.model_create_multi
     def create(self, vals_list):
         # We generate a standard reference
         for vals in vals_list:
-            vals['name'] = self.env['ir.sequence'].next_by_code('repair.order') or '/'
+            # vals['name'] = self.env['ir.sequence'].next_by_code('repair.order') or '/'
+            if not vals.get('name', False) or vals['name'] == _('New'):
+                picking_type_id = vals.get('picking_type_id')
+                if not picking_type_id:
+                    picking_type_id = self._get_default_picking_type_id(vals.get('company_id', self.env.company.id))
+                    vals['picking_type_id'] = picking_type_id
+                vals['name'] = self.env['stock.picking.type'].browse(picking_type_id).sequence_id.next_by_id()
+            if not vals.get('procurement_group_id'):
+                vals['procurement_group_id'] = self.env["procurement.group"].create({'name': vals['name']}).id
         return super().create(vals_list)
 
+    def write(self, vals):
+        super().write(vals)
+        for order in self:
+            if 'picking_type_id' in vals:
+                order.operation_move_ids.write({'picking_type_id': order.picking_type_id.id})
+
+            moves = order.operation_move_ids.filtered('sale_line_id')
+            if 'under_warranty' in vals:
+                if order.under_warranty:
+                    moves.mapped('sale_line_id').write({'price_unit' : 0.0})
+                else:
+                    moves.mapped('sale_line_id')._compute_price_unit()
+
+            if 'state' in vals and order.state == 'done':
+                for move in moves:
+                    move.sale_line_id.write({'qty_delivered' : move.sale_line_id.product_uom_qty})
```

Comment:
I would call a method _action_done in this case to match the button action and the state click